packages/components/src/utils/react.ts

```diff
@@ -0,0 +1,244 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	useRef,
+	useInsertionEffect,
+	useCallback,
+	useEffect,
+	useState,
+} from '@wordpress/element';
+
+/**
+ * Any function.
+ */
+export type AnyFunction = ( ...args: any ) => any;
+
+/**
+ * Creates a stable callback function that has access to the latest state and
+ * can be used within event handlers and effect callbacks. Throws when used in
+ * the render phase.
+ *
+ * @example
+ *
+ * ```tsx
+ * function Component(props) {
+ *   const onClick = useEvent(props.onClick);
+ *   React.useEffect(() => {}, [onClick]);
+ * }
+ * ```
+ */
+export function useEvent< T extends AnyFunction >( callback?: T ) {
+	const ref = useRef< AnyFunction | undefined >( () => {
+		throw new Error( 'Cannot call an event handler while rendering.' );
+	} );
+	useInsertionEffect( () => {
+		ref.current = callback;
+	} );
+	return useCallback< AnyFunction >(
+		( ...args ) => ref.current?.( ...args ),
+		[]
+	) as T;
+}
+
+/**
+ * `useResizeObserver` options.
+ */
+export type UseResizeObserverOptions = {
+	/**
+	 * Whether to trigger the callback when an element's ResizeObserver is
+	 * first set up.
+	 *
+	 * @default true
+	 */
+	fireOnObserve?: boolean;
+};
+
+/**
+ * Fires `onResize` when the target element is resized.
+ *
+ * **The element must not be stored in a ref**, else it won't be observed
+ * or updated. Instead, it should be stored in a React state or equivalent.
+ *
+ * It sets up a `ResizeObserver` that tracks the element under the hood. The
+ * target element can be changed dynamically, and the observer will be
+ * updated accordingly.
+ *
+ * By default, `onResize` is called when the observer is set up, in addition
+ * to when the element is resized. This behavior can be disabled with the
+ * `fireOnObserve` option.
+ *
+ * @example
+ *
+ * ```tsx
+ * const [ targetElement, setTargetElement ] = useState< HTMLElement | null >();
+ *
+ * useResizeObserver( targetElement, ( element ) => {
+ *   console.log( 'Element resized:', element );
+ * } );
+ *
+ * <div ref={ setTargetElement } />;
+ * ```
+ */
+export function useResizeObserver(
```

Comment:
Why do we have this hook here and useResizeObserver in compose as well?

Reply:
I'll let @DaniGuardiola reply with more details if necessary, but from earlier conversations he took a look at the `useResizeObserver` hook in `@wordpress/compose` and found it overly complex and difficult to compose with other smaller utilities (like in this PR).

Now that resize observers are well supported by browsers, we could look at simplifying the `@wordpress/compose` version with something similar to the above implementation

Reply:
I understand that it's easier to duplicate stuff and add new internal utils but It would be good to avoid recreating the same things over and over. A small effort can bring benefits to the whole project. In fact I had started improving it in the past but lost interest on it. https://github.com/WordPress/gutenberg/pull/41001

Reply:
The hook called `useResizeObserver` has a misleading name. It does use a `ResizeObserver` under the hood, but has a hyper-specific API focused on a narrow use-case (getting the viewport-relative size of a parent element). Neither the API nor the sizing computation worked in my case. The existing hook should probably be called `useParentElementBoundingClientRect`, and even then it would not describe all of the nuance fully because it's a chunky hook full of undocumented logic.

My hook here is completely different and does a single thing: takes an element and a callback, and executes the callback on resize by using a ResizeObserver. Nothing more. All of the nuance is also clearly documented. As you can see further down this file, it is nicely composed with other utilities (like one to obtain the sizing and position in the appropriate context here) to achieve what was needed for this task.

So to be clear: there isn't any prior code that does what this set of utilities do, so there's no duplication. In order to effectively avoid duplication in a future-proof manner, utilities must be carefully crafted to do a single thing with a minimal API, so that they can be composed with other utilities. Unfortunately, that's not the case of the pre-existing `useResizeObserver` utility.

Reply:
I wrote more specifics here: https://github.com/WordPress/gutenberg/pull/60560#discussion_r1610163431

Reply:
If you're interested, I'd be happy to pair on bringing atomic-sized and composable utilities for resize observer related stuff to the shared utils package. My hook here for instance is a good primitive to build things upon, since it doesn't get simpler than that. We can easily build more specific variants from it, like the one present in this file that tracks the overflow-parent-relative rect. It should also be quite easy to add more ergonomic APIs to work with refs more directly. Just ping me if you're interested :)

Reply:
It is less about how the hook works and more about having one way to do something in the whole repo and be consistent and avoid silos (components doing things one way, the rest doing something else), I think it's an important principle that we should stick to. 

Now on the hook, both hooks have the same purpose, observe the changes of size of an element. If you check the PR that I linked, it actually offers a very similar API to the hook in components, so it's possible to migrate (offer both while allowing migration).

Also, on my opinion, some changes are not worth it. I'm not saying it's the case here, I don't know but sometimes it's ok to stick for something that is a bit older and maybe not ideal of the shiniest approach requires big refactoring without important gains (either performance or else).

Reply:
I appreciate the points you're making.

That said, I suggest you look at the links I dropped. As I explained, the existing hook does a very different thing to what I do, in a very different way. It wouldn't have worked for my use case. Also, I would argue that the fact that a utility abstracts a web primitive shouldn't block the use of said primitive in other places where it'd be useful but the abstraction just doesn't work, like here.

I understand that there's a history and a context to all of this, don't think I'm ignoring this fact - I still think it's productive to recognize the limitations of it and understand why it's not ideal, design-wise. Independently of what the pragmatic approach is here, I'm arguing that writing utilities in the style I've described (and used in this PR) is a more effective way to reduce duplication and future-proof them. For example, you could very easily construct the older hook using mine, but it'd be outright impossible the other way around. I think composition is king when it comes to these things.

I think there's a lot to learn from the ["do one thing and do it well" principle of "Unix philosophy"](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well).

Reply:
Upon closer inspection, I do believe my hook could probably be renamed to something like `useOnElementSizeUpdate`, since the point of it is to just be notified of size updates and the resize observer is merely an implementation detail. In fact, the [ResizeObserverEntry](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry) objects are ignored by it (they are useless for this use-case), and there's a special behavior by default that fires the handler initially with the purpose of keeping things up to date always.

Maybe this clarifies why the other hook is not a good fit here better?

I also took a closer look at your PR and I think it's much closer to what I would call a minimal and composable API, love it! Definitely much better than the absolutely positioned child approach, IMO.

Reply:
Alright so, since I've been getting a bit of pushback on this, I've stopped to do a bit of a reset and look at this from the outside, and I've realized a few things. I think I have been explaining myself poorly, and grown a bit frustrated which has led me to missing some important things. Let me elaborate:

- I think my main mistake has been that I haven't explained the initial purpose of this logic. Initially, it didn't even have a `ResizeObserver`. In most cases, the observer won't even fire, ever! I assume this'll be surprising because I have failed to explain myself properly. Let me recap:
  - I needed to track which tab is active, and then obtain its position and size. That was easy enough to achieve by wiring a bit of Ariakit state with some logic to compute the size (relative to the offset parent, long story, yadda yadda).
  - However, in some rare cases, a tab would change its size in a re-render (e.g. in the block editor, left sidebar, the first tab initially has a label of "Document", which then gets instantly replaced by "Template", or "Post", or "Page".
  - Since the size computation was already done by that re-render, the indicator would remain with a size that was computed with "Document", which does not fit the new tab with the new label.
  - To solve for this, I needed a way to be notified when this happens to re-compute size and position. For that, a `ResizeObserver` (at least the part where it fires a callback when there's a resize) is the perfect tool.
  - Since I was using it, someone suggested that I used the pre-existing hook. However, for the reasons stated (and a few more), it just doesn't fit the use-case, and it wouldn't have any benefit over just using the primitive directly, as I do.
- Another mistake that I've realized now is that my hook here is not named appropriately. It should probably be called something like `useTrackElementRectUpdates` instead. The main use-case, as explained, is to be notified when the target element is swapped (e.g. from tab 1 to tab 2) to obtain an up-to-date size and position, **and** also get those updates when the same target element alters those properties. It leaves the implementation of actually computing any values to the user, which allows for the neat composition I achieve below in this file. Again, `ResizeObserver` is purely an implementation detail, and it could be very easily swapped by something else in the future if a better browser API were to come out.
- I also think I overlooked what the old hook is doing. This was also a big mistake, as I didn't realize that it was much more pegged to the `ResizeObserver` API than I thought. I have now looked at it and, after learning more about [ResizeObserverEntry](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry) objects (which I have never had a use-case for), I realize that I've been obtusely arguing that the hook has too much custom logic, which is false. I apologize for this. I still have some opinions on the API (as you do, for example the change in your PR is much more aligned with what I would describe as a better API for this).

All in all, I've had to do some self-reflection here and I apologize for being obtuse, and for overlooking important details. I let some mild frustration cloud my judgment. My bad!

Reply:
TL;DR: sorry for being a dummy dum dum and thank you for engaging with me :)

With this new perspective, I'd be happy to re-open your PR and help get it to land @youknowriad

Reply:
Thanks for the discussion here. Just want to say that this discussion doesn't have to be a blocker for the current PR. I'm just trying to share my mental model and some guiding principles.

Reply:
Thank you Dani for sharing in detail your train of thoughts, it helps a lot! I personally empathise with Riad's sentiment — consistency is important in a project like Gutenberg. And if we could improve the `useResizeObserver` hook from `@wordpress/compose`, the whole project (and its third-party consumers) would benefit from it.

As for the next steps, we could do the following:

1. It would great to pick up Riad's work from https://github.com/WordPress/gutenberg/pull/41001, and see if we can provide a new version of the hook based on modern Web technologies that satisfies the requirements of its consumers in the editor, providing a good, non-breaking migration strategy
2. In parallel, we can continue moving utils to a separate folder, private to the components package. For the "local" `useResizeObserver` utility discussed in this PR, we can add a TODO comment and remove it in favor of the `@wordpress/compose` version when we're done with the rewrite (see above point).

What do y'all think?

Reply:
@ciampo agreed on 1! for 2, unfortunately the hook will still not work for the reasons I stated in my first bullet point. I need an event to fire when the target element is first observed, both initially and subsequently, and I also don't need the overhead of the resize observer entries.

I think I should just rename the hook so that it can't be confused as an adapter for ResizeObserver (which it isn't), for which we already have one. It's a utility that does something, and happens to use the `ResizeObserver` browser API under the hood, as an implementation detail. Hopefully this makes sense.

Reply:
Sounds good to me, at least for the short term.

As we work on refactoring the `useResizeObserver` shared utility from `@wordpress/compose` we could also evaluate adding new features to it, like firing when the target element is first observed.

Reply:
I updated the hooks name to `useTrackElementRectUpdates` for clarity, and updated some of the docs. In the process of doing that, I discovered something interesting to consider in future ResizeObserver-related work: https://stackoverflow.com/a/60026394

Reply:
A few updates here:

- https://github.com/WordPress/gutenberg/pull/64820
- https://github.com/WordPress/gutenberg/pull/64926
- https://github.com/WordPress/gutenberg/pull/64943